src/auth/src/hash.rs

```diff
@@ -0,0 +1,227 @@
+pub fn scram256_verify(password: &String, hashed_password: &str) -> Result<(), VerifyError> {
+    let opts = scram256_parse_opts(hashed_password)?;
+    let hashed = hash_password_with_opts(&opts, password);
+    let scram = scram256_hash_inner(hashed);
+    if *hashed_password == scram.to_string() {
+        Ok(())
+    } else {
+        Err(VerifyError::InvalidPassword)
+    }
+}
+
+/// Parses a SCRAM-SHA-256 hash and returns the options used to create it.
+fn scram256_parse_opts(hashed_password: &str) -> Result<HashOpts, VerifyError> {
+    let parts: Vec<&str> = hashed_password.split('$').collect();
+    if parts.len() != 3 {
+        return Err(VerifyError::MalformedHash);
+    }
+    let scheme = parts[0];
+    if scheme != "SCRAM-SHA-256" {
+        return Err(VerifyError::MalformedHash);
+    }
+    let auth_info = parts[1].split(':').collect::<Vec<&str>>();
+    if auth_info.len() != 2 {
+        return Err(VerifyError::MalformedHash);
+    }
+    let auth_value = parts[2].split(':').collect::<Vec<&str>>();
+    if auth_value.len() != 2 {
+        return Err(VerifyError::MalformedHash);
+    }
+
+    let iterations = auth_info[0]
+        .parse::<u32>()
+        .map_err(|_| VerifyError::MalformedHash)?;
+
+    let salt = BASE64_STANDARD
+        .decode(auth_info[1])
+        .map_err(|_| VerifyError::MalformedHash)?;
+
+    let salt = salt.try_into().map_err(|_| VerifyError::MalformedHash)?;
+
+    Ok(HashOpts {
+        iterations: NonZeroU32::new(iterations).ok_or(VerifyError::MalformedHash)?,
+        salt,
+    })
+}
+
+/// The SCRAM-SHA-256 hash
+struct ScramSha256Hash {
+    /// The number of iterations used for hashing
+    iterations: NonZeroU32,
+    /// The salt used for hashing
+    salt: [u8; 32],
+    /// The server key
+    server_key: [u8; SHA256_OUTPUT_LEN],
+    /// The client key
+    client_key: [u8; SHA256_OUTPUT_LEN],
+}
+
+impl Display for ScramSha256Hash {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "SCRAM-SHA-256${}:{}${}:{}",
+            self.iterations,
+            BASE64_STANDARD.encode(&self.salt),
+            BASE64_STANDARD.encode(&self.client_key),
+            BASE64_STANDARD.encode(&self.server_key)
+        )
+    }
+}
+
+fn scram256_hash_inner(hashed_password: PasswordHash) -> ScramSha256Hash {
+    let signing_key = Key::new(HMAC_SHA256, &hashed_password.hash);
+    let client_key = hmac::sign(&signing_key, b"Client Key");
+    let server_key = hmac::sign(&signing_key, b"Server Key");
+
+    ScramSha256Hash {
+        iterations: hashed_password.iterations,
+        salt: hashed_password.salt,
+        server_key: server_key.as_ref().try_into().unwrap(),
+        client_key: client_key.as_ref().try_into().unwrap(),
+    }
+}
+
+fn hash_password_inner(opts: &HashOpts, password: &[u8]) -> [u8; SHA256_OUTPUT_LEN] {
+    let mut salted_password = [0u8; SHA256_OUTPUT_LEN];
+    pbkdf2::derive(
```

Comment:
Not technically constant time, but I don't think we are really subject to timing attacks here

Reply:
is there another document where we outline the set of attacks we are defending against? I can't see anything in [the design](https://github.com/MaterializeInc/materialize/pull/32005/files?short_path=076932b#diff-076932baba2624ac63805005038b0708b0fe4973ca680047cde716a96bdc74a0)